### PR TITLE
ENH: remove the need for a working directory & other cleanups

### LINKED ISSUE #317: BUG: install_subdir() in meson.build does not honor exlude_files/directories
I'm trying to use `install_subdir()` to simplify the inclusion of Python source files and other data files in the final wheel.  However, the `install_subdir()` function is not honoring the  `exlude_files/directories` options. 

Core Meson is installing things properly and exclusions are honored. However, when `mesonpy` takes over, it ends up copying the subtree from the source directory, and not the install directory as I was expecting. That's the reason the exclusions are not honored: things are being copied from the wrong location. As Meson is all about out-of-tree builds, I would argue this is a bug.

The following trivial patch fixes the problem:
```diff
diff --git a/mesonpy/__init__.py b/mesonpy/__init__.py
index 40b9645..fe5055d 100644
--- a/mesonpy/__init__.py
+++ b/mesonpy/__init__.py
@@ -486,7 +486,7 @@ class _WheelBuilder():
                 )
                 if install_details:
                     scheme, destination = install_details
-                    wheel_files[scheme].append((destination, file))
+                    wheel_files[scheme].append((destination, copy_files[file]))
                     continue
                 # not found
                 warnings.warn(
```
However, this patch makes some tests in `tests/test_wheel.py` break.
I'm not familiar enough with Meson and meson-python to triage these test failures. Is anyone able to help?

### LINKED ISSUE #355: Prevent users from overriding our required arguments
We pass several arguments to Meson calls, some of them are required, some of them are simply meant as defaults, which the user can override (see GH-329). We should pass the our defaults first, then the user arguments, and finally, the arguments we require to not be changed.

Supersedes GH-330

### PR DESCRIPTION
**Pull request by dnicolodi**
This brings back the simplifications I proposed in #340 in a way compatible with having the wheel generated from files installed by `meson install` in a temporary destination directory. This covers a very minimal part of the work done in #348 too. It should be trivial to rebase #348 on top of this.

Fixes #355 and #330 as there are no arguments the user can supply that would interfere with meson-python internals. It also fixes #317 in the same way as #348 does.

### CODE REVIEW

**Comment by dnicolodi** [comment]
The macOS test failure are the same found in #348 and will have the same solution.

**Comment by rgommers** [comment]
Thanks @dnicolodi. Trying to think of a merge strategy here - I think the status is:
- this seems to overlap with, and improve on, the first 3 commits of gh-348.
- this is blocked on the macOS RPATH test failures right now, which are addressed by ``gh-348`` (and that's the key tricky and unfinished thing in that PR). This PR could temporarily xfail them though?
- ``gh-348`` also has tests for `subdir(..., exclude_files)` that are useful and will pass in combination with this PR. But there's no issue in adding those later.

**Comment by dnicolodi** [comment]
I see this more like preparation work for landing #348 with the resulting code being in better shape (if we agree that the resulting code after merging this PR is "better").

From a "build things from the ground up" point of view, I would merge this and rebase the fixes in #348 on top. Or we can do the other way around and have this merged after #348 to cleanup the implementation. I think merging this first results in less (or easier to resolve) rebase conflicts and will make looking the commit history easier, but we can go either way. This PR being ready to be merged, while the proper solution for the rpath issues in #348 are still being discussed, is one reason to have this merged first. If you want to merge with all green tests, I can mark the rpath test as `xfail`.

I can take care of rebasing #348 on top of this.

The tests for `exclude_files` probably found a bug in Meson on Windows, thus these are also marked as `xfail` on that platform. I started investigating that on the Meson side, but I didn't reach a conclusion yet.

**Comment by dnicolodi** [comment]
PR #348 rebased on top of this is here https://github.com/dnicolodi/meson-python/tree/pr348

**Comment by FFY00** [comment]
I removed the v0.13.0 milestone, as this getting in is not critical for it, and we should release 0.13.0 as soon as we are able to. If this is in shape to get it by the time we sort the release blockers, great, otherwise we can just merge it afterwards and have it in 0.14.0.

**Comment by dnicolodi** [comment]
This is ready to be merged since a long time. It just needs a review. I don't understand why you think otherwise.

**Comment by FFY00** [comment]
It conflicts with #348, which needs to get in for the new release.

**Comment by dnicolodi** [comment]
I already posted the conflict resolution.

**Comment by FFY00** [comment]
But the PR may get more changes before merging. If you have an updated PR by then, great, we can merge it, otherwise we just do it for the next release.

**Comment by dnicolodi** [comment]
What's the problem with merging this now, rebase #348 on top, and take it from there?

**Comment by FFY00** [comment]
It can unnecessarily delay the release. Like I said, this is not critical, unlike the other PR, so I am not merging this now.

**Comment by FFY00** [comment]
This PR hasn't even been reviewed btw, so it might still require changes, etc. I am prioritizing the release blockers.

**Comment by FFY00** [comment]
There are still unaddressed items in that PR that I need to deal with, and I haven't even pushed any changes from the feedback received. When I do not have any more actionable items there, I will review this.

**Comment by dnicolodi** [comment]
@rgommers I left out the commits with behavior changes from this PR. This is now just refactoring and a change in the structure of temporary directories. It would be nice to have this merged as a first step toward fixing the linked issues.

**Review by rgommers — APPROVED** [review]
Thanks @dnicolodi. Restructuring this to have no behavior changes was helpful. Sorry it took so long to review. I verified this as well as I could, and tested with SciPy - it all looks good. The code is easier to understand after this - and the simplifications in the tests are a good sign too. So in it goes.

**Comment by rgommers** [comment]
@dnicolodi WDYT about tagging a 0.14.0 release? The PR queue is pretty much empty, so now would be a good time. We can quickly do a 0.15.0 too in case the subsequent PR to this one arrives quickly.

**Comment by dnicolodi** [comment]
> WDYT about tagging a 0.14.0 release?

I don't know how critical it is to push a release with the fixes that already landed in `main`. There are three known bugs related to the fact that meson-python takes the files for the wheel from the source directory instead than from the build directory: (1) the handling of `exclude_path` and `exclude_directory` for `install_subdir`, and (2) removing the RPATH pointing to the build directory. In addition, there is a bug (3) in how sub direcrtories of direcotries installed with `install_subdir` are handled. I have patches for all three, thus getting these fixes in should be fast. However, investigating (2) I found a bug in Meson, thus having correct RPATH handling would require a new Meson release.

**Comment by rgommers** [comment]
It's not critical to do it right this minute, so let's wait for at least the fixes for the bugs that don't depend on a new Meson release?

I'm also going through the backlog now and see what can be closed or easily addressed. Some of the documentation issues are straightforward, I'll try to fix a couple after dinner.

**Pull request merged**
